pkg/imgpkg/imageset/image_set.go

```diff
@@ -77,57 +79,135 @@ func (o *ImageSet) Import(imgOrIndexes []imagedesc.ImageOrIndex,
 		item := item // copy
 
 		go func() {
 			importThrottle.Take()
 			defer importThrottle.Done()
+			addImageOrImageIndexToTaggableMapForMultiWrite(item, errCh, importRepo, registry, imageOrIndexesToWrite, imageOrIndexesToWriteLock)
+		}()
+	}
 
-			existingRef, err := regname.NewDigest(item.Ref())
-			if err != nil {
-				errCh <- err
-				return
-			}
+	err := checkForAnyAsyncErrors(imgOrIndexes, errCh)
+	if err != nil {
+		return nil, err
+	}
 
-			importDigestRef, err := o.importImage(item, existingRef, importRepo, registry)
-			if err != nil {
-				errCh <- fmt.Errorf("Importing image %s: %s", existingRef.Name(), err)
-				return
-			}
+	err = registry.MultiWrite(imageOrIndexesToWrite, o.concurrency)
+	if err != nil {
+		return nil, err
+	}
 
-			var regImage regv1.Image
-			if item.Image != nil {
-				regImage = *item.Image
-			}
-			var regImageIndex regv1.ImageIndex
-			if item.Index != nil {
-				regImageIndex = *item.Index
-			}
-			importedImages.Add(ProcessedImage{
-				UnprocessedImageRef: UnprocessedImageRef{existingRef.Name(), item.Tag()},
-				DigestRef:           importDigestRef.Name(),
-				Image:               regImage,
-				ImageIndex:          regImageIndex,
-			})
-			errCh <- nil
+	errChVerifyImages := make(chan error, len(imgOrIndexes))
+	for _, item := range imgOrIndexes {
+		item := item // copy
+
+		go func() {
+			importThrottle.Take()
+			defer importThrottle.Done()
+
+			o.verifyProcessedImages(item, errChVerifyImages, importRepo, registry, importedImages)
 		}()
 	}
 
+	err = checkForAnyAsyncErrors(imgOrIndexes, errChVerifyImages)
+	if err != nil {
+		return nil, err
+	}
+
+	return importedImages, nil
+}
+
+func checkForAnyAsyncErrors(imgOrIndexes []imagedesc.ImageOrIndex, errCh chan error) error {
 	for i := 0; i < len(imgOrIndexes); i++ {
 		err := <-errCh
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
+	return nil
+}
 
-	return importedImages, nil
+func addImageOrImageIndexToTaggableMapForMultiWrite(item imagedesc.ImageOrIndex, errCh chan error, importRepo regname.Repository, registry ImagesReaderWriter, imageOrIndexesToWrite map[regname.Reference]regremote.Taggable, lock *sync.Mutex) {
+	itemDigest, err := item.Digest()
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	tag := fmt.Sprintf("imgpkg-%s-%s", itemDigest.Algorithm, itemDigest.Hex)
+	uploadTagRef, err := regname.NewTag(fmt.Sprintf("%s:%s", importRepo.Name(), tag))
+	if err != nil {
+		errCh <- fmt.Errorf("Building upload tag image ref: %s", err)
+		return
+	}
+
+	switch {
+	case item.Image != nil:
+		itemRef, err := regname.ParseReference(item.Ref())
+		if err != nil {
+			errCh <- fmt.Errorf("Unable to parse reference: %s: %s", item.Ref(), err)
+			return
+		}
+
+		imageToWrite := regv1.Image(*item.Image)
+
+		if imageBlobsCanBeMounted(itemRef, uploadTagRef) {
+			imageToWrite, err = item.MountableImage(registry)
+			if err != nil {
+				imageToWrite = regv1.Image(*item.Image)
+			}
+		}
+
+		lock.Lock()
```

Comment:
i think it would be better for this function to return uploadTagRef and taggable image/index, and have imageOrIndexesToWrite and associated lock be used in the same scope as it's defined. (this way you dont have to pass them around) to mutate.

Reply:
(goes hand in hand with not passing down channel and dealing with them in the same scope as go routine).

Reply:
I can make this change, although it's not obvious to me why it is 'better'